### ISSUE TITLE
Concurrent RPCs over TLS with nodejs client fail due to TLS errors

### ISSUE THREAD

**Issue by shimos** [grpc/grpc#12642]
### What version of gRPC and what language are you using?

- grpc 1.6.0

### What operating system (Linux, Windows, …) and version?

- Ubuntu Linux 16.04.3 LTS
- Kernel: Linux 4.4.0-1022-aws #31-Ubuntu (x86_64)
- OpenSSL 1.0.2g

### What runtime / compiler are you using (e.g. python version or version of gcc)

- node v6.11.2

### What did you do?

We ran a simple grpc server and client, both written in nodejs, on the same host. (See below for codes)
The client repeatedly invoked RPCs concurrently to the same server, and we increased the number of the simultaneous requests.

Server source code:
~~~~
'use strict';

process.env.GRPC_SSL_CIPHER_SUITES = "ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384";

const grpc = require('grpc');
const crypto = require('crypto');
const fs = require('fs');

const testProto = grpc.load(__dirname + '/test.proto');
const test = testProto.test;

const server = new grpc.Server();
server.addService(test.TestService.service, { GetHashedValue: getHashedValue });

const rootCA = fs.readFileSync(__dirname + '/tls/ca.crt');
const serverpem = fs.readFileSync(__dirname + '/tls/server.crt');
const privkey = fs.readFileSync(__dirname + '/tls/server.key');
const cred = grpc.ServerCredentials.createSsl(rootCA, [{ private_key : privkey, cert_chain : serverpem }], false);

server.bind('0.0.0.0:9998', cred);
server.start();

function getHashedValue(call, callback) {
	callback(null, __getHashedValue(call.request));
}

function __getHashedValue(input) {
	// Dummy calculation
	var i = 0;
	while (!crypto.createHash('sha256').update(input.data + i).digest('hex').startsWith('00')) {
		i++;
	}
	
	return {
		hash_md5 : crypto.createHash('md5').update(input.data).digest('hex'),
		hash_sha256 : crypto.createHash('sha256').update(input.data).digest('hex'),
		nonce : i.toString()
	};
}
~~~~

Client source code (`mutliclient.js`)
~~~~
'use strict';

const grpc = require('grpc');
const fs = require('fs');
const util = require('util');

const testProto = grpc.load(__dirname + '/test.proto');
const test = testProto.test;

process.env.GRPC_SSL_CIPHER_SUITES = "ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384";

var opts = { 'grpc.ssl_target_name_override' : 'orderer0.example.com',
             'grpc.default_authority' : 'orderer0.example.com' };

function GetHashedValue(stub, data) {
	return new Promise((resolve, reject) => {
		let t = setTimeout(() => { reject('timeout'); }, 10000);
		stub.getHashedValue({ data : data }, (err, output) => {
			clearTimeout(t);
			if (err) {
				reject(err);
			} else {
				resolve(output);
			}
		});
	});
}

function getStub(addr) {
	const rootCA = fs.readFileSync(__dirname + '/tls/ca.crt');
	const cred = grpc.credentials.createSsl(rootCA);
	
	return new test.TestService(addr, cred, opts);
}

var mainPromise = Promise.resolve();
for (let i = 0; i < 1000; i++) {
	mainPromise = mainPromise.then(() => {
		let promises = [];
		for (let c = 0; c < 10; c++) {
			promises.push(GetHashedValue(getStub(util.format('localhost:%d', 9998)), 'abcde')
			              .then(
				              (output) => {
				              },
				              (err) => {
					              console.log(util.format("Error : i = %d, c = %d", i, c));
					              console.log(err);
				              }
			              ));
		}
		return Promise.all(promises);
	});
}
~~~~

### What did you expect to see?

All RPCs finished successfully with no error message.

### What did you see instead?

SSL-related error messages in the client side as below:

```
(10 simultaneous requests for 1000 loops)
$ node multiclient.js
E0918 19:15:13.549735953    9191 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 19:15:13.549784041    9191 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 19:15:13.549809507    9191 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 19:15:13.549817436    9191 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 19:15:13.549823886    9191 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
Error : i = 298, c = 9
timeout
E0918 19:15:23.555298724    9191 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 19:15:23.555327694    9191 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 19:15:23.555343490    9191 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 19:15:23.555348706    9191 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 19:15:23.555352415    9191 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
E0918 19:15:23.555987397    9191 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 19:15:23.556007633    9191 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 19:15:23.556021100    9191 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 19:15:23.556024783    9191 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 19:15:23.556027500    9191 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
E0918 19:15:23.556827941    9191 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 19:15:23.556849317    9191 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 19:15:23.556863770    9191 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 19:15:23.556867610    9191 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 19:15:23.556870305    9191 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
E0918 19:15:23.557518465    9191 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 19:15:23.557534417    9191 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 19:15:23.557547464    9191 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 19:15:23.557551147    9191 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 19:15:23.557553896    9191 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
E0918 19:15:23.558236543    9191 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 19:15:23.558256848    9191 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 19:15:23.558271198    9191 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 19:15:23.558275183    9191 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 19:15:23.558277977    9191 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
E0918 19:15:23.558943764    9191 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 19:15:23.558964145    9191 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 19:15:23.558977869    9191 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 19:15:23.558981556    9191 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 19:15:23.558984202    9191 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
Error : i = 299, c = 0
timeout
Error : i = 299, c = 1
timeout
Error : i = 299, c = 2
timeout
Error : i = 299, c = 3
timeout
Error : i = 299, c = 4
timeout
Error : i = 299, c = 5
timeout
E0918 19:15:36.283131370    9191 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 19:15:36.283162646    9191 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 19:15:36.283180800    9191 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 19:15:36.283188059    9191 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 19:15:36.283194759    9191 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
Error : i = 353, c = 9
timeout
E0918 19:15:46.281421324    9191 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 19:15:46.281450162    9191 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 19:15:46.281465908    9191 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 19:15:46.281470151    9191 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 19:15:46.281473335    9191 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
Error : i = 354, c = 0
timeout
```

No error messages are shown in the server.

### Anything else we should know about your project / environment?

The error occurred probabilistically (so looks like due to some timing bug), and we observed that this can happen even with a small number of concurrency(such as c = 2).

We have got another type of error message like:
```
E0918 16:31:23.879222619    8681 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 16:31:23.879252246    8681 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 16:31:23.879300087    8681 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 16:31:23.879306545    8681 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 16:31:23.879309814    8681 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
E0918 16:31:23.880015758    8681 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 16:31:23.880037639    8681 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 16:31:23.880055859    8681 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 16:31:23.880063476    8681 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 16:31:23.880069342    8681 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
E0918 16:31:23.880760460    8681 ssl_transport_security.c:575] Could not add root certificate to ssl context.
E0918 16:31:23.880782242    8681 ssl_transport_security.c:1297] Cannot load server root certificates.
E0918 16:31:23.880801987    8681 security_connector.c:857]   Handshaker factory creation failed with TSI_INTERNAL_ERROR.
E0918 16:31:23.880809411    8681 secure_channel_create.c:112] Failed to create secure subchannel for secure name 'localhost:9998'
E0918 16:31:23.880815629    8681 secure_channel_create.c:143] Failed to create subchannel arguments during subchannel creation.
E0918 16:31:23.881436947    8681 ssl_transport_security.c:427] Corruption detected.
E0918 16:31:23.881468099    8681 ssl_transport_security.c:403] error:0B07C065:x509 certificate routines:X509_STORE_add_cert:cert already in hash table
E0918 16:31:23.881476868    8681 ssl_transport_security.c:403] error:0B07C065:x509 certificate routines:X509_STORE_add_cert:cert already in hash table
E0918 16:31:23.881484395    8681 ssl_transport_security.c:403] error:0B07C065:x509 certificate routines:X509_STORE_add_cert:cert already in hash table
E0918 16:31:23.881491065    8681 secure_endpoint.c:177]      Decryption error: TSI_DATA_CORRUPTED
Error : i = 514, c = 7
timeout
Error : i = 514, c = 8
timeout
Error : i = 514, c = 9
timeout
```

**Comment by murgatroid99** [comment]
@jboeuf Do you recognize this issue?

**Comment by jboeuf** [comment]
@justinburke FYI.

No, I've never seen that before.

**Comment by justinburke** [comment]
If this ends up being difficult to reproduce on the side of gRPC team, would it be possible to get a Dockerfile that demonstrates the behavior? I know that might be a big request, but might be a good way to make forward progress here.

**Comment by shimos** [comment]
@justinburke Thank you. Sure, I will create and attach it later.

**Comment by shimos** [comment]
Attached a Docker image which reproduced the issue in my AWS environment (size is t2.large) : [grpc-issue-12642.zip](https://github.com/grpc/grpc/files/1330862/grpc-issue-12642.zip)
As it looks like timing problem, you might have to repeat it several times to reproduce it.

**Comment by shimos** [comment]
Any updates?

I tested this in some older versions of gRPC.
- With 1.3.2, it produced the same errors as 1.6.0.
- With 1.2.4, since the combination of nodejs server and client made the system almost stuck apparently due to memory leak, I ran go server and nodejs client. Then, the client got timeout and had some different error messages.
```
E1002 11:46:11.767082508   20576 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
E1002 11:46:11.767363417   20576 ssl_transport_security.c:439] SSL_read returned 0 unexpectedly.
E1002 11:46:11.767373474   20576 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
E1002 11:46:11.767391498   20576 ssl_transport_security.c:439] SSL_read returned 0 unexpectedly.
E1002 11:46:11.767398565   20576 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
E1002 11:46:11.767415451   20576 ssl_transport_security.c:439] SSL_read returned 0 unexpectedly.
E1002 11:46:11.767422255   20576 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
E1002 11:46:11.767438777   20576 ssl_transport_security.c:439] SSL_read returned 0 unexpectedly.
E1002 11:46:11.767445652   20576 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
E1002 11:46:11.767464311   20576 ssl_transport_security.c:439] SSL_read returned 0 unexpectedly.
E1002 11:46:11.767471150   20576 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
(repeated many times)
```

Hope this might help.

**Comment by davidkel** [comment]
Is there any outlook for this issue ?

**Comment by WandyLau** [comment]
Hi,
I have a look at the source code.  As the log suggests, the root cause may come from boringssl.
And I found an improper logic here in ssl_transport_security.c in function 
X509_STORE_add_cert. ==> X509_OBJECT_retrieve_match ==> x509_object_cmp, where may be the issue. The code is as below:

        for (i = idx; i < sk_X509_OBJECT_num(h); i++) {                                                                                                 
    588        obj = sk_X509_OBJECT_value(h, i);                                                                                                           
    589        if (x509_object_cmp                                                                                                                   
    590            ((const X509_OBJECT **)&obj, (const X509_OBJECT **)&x))                                                                                 
    591            return NULL;                                                                                                                            
    592        if (x->type == X509_LU_X509) {                                                                                                              
    593            if (!X509_cmp(obj->data.x509, x->data.x509))                                                                                            
    594                return obj;                                                                                                                         
    595        } else if (x->type == X509_LU_CRL) {                                                                                                        
    596            if (!X509_CRL_match(obj->data.crl, x->data.crl))                                                                                        
    597                return obj;                                                                                                                         
    598        } else                                                                                                                                      
    599            return obj;                                                                                                                             
    600    }                      
    
The first glance, is any wrong in line 589 and 591?  just a clue.

**Comment by WandyLau** [comment]
As I read the source code, the error originate from the function **X509_STORE_add_cert**, I will try to debug this. But I am new to the grpc development. If anyone know how to debug it, please give me some advice.

**Comment by harrisob** [comment]
Any luck with this one ?

**Comment by WandyLau** [comment]
It seems that I can resolve the ssl problem, but the timeout still appears. I will push my patch then. I am not sure if the modification is right. But I have never seen the ssl error then. Just let the maintainer tell.

**Comment by WandyLau** [comment]
The above is not the solution after my test.  But I think I found the root cause with the docker image provided by @shimos 
This is just a trick. When I print out the credentials out, there are so many random characters appending. The length and the content are all random. And sometimes it is the part of the ca root. This will raise the ssl issue.
And If the credential is too big with so much random content, i think it is also the root cause of the time out issue. 

That means the credentials got too much from the memory, this is maybe a security issue too. I just add string
terminator to the credentials. If anyone got more elegant fix, just do it. https://github.com/grpc/grpc-node/compare/master...WandyLau:patch-1

**Comment by WandyLau** [comment]
Well this is also a security problem, as the client side will read more data than expected from the memory.

**Comment by eastlondoner** [comment]
I'm getting this in python as well (presumably using the same underlying C library).
using a multiprocessing pool to send grpc requests. If pool size is 1 everything is fine. If pool size > 1 then:

`E1124 11:01:22.214234154   10566 ssl_transport_security.c:435] Corruption detected.
E1124 11:01:22.214365388   10566 ssl_transport_security.c:411] error:100003fc:SSL routines:OPENSSL_internal:SSLV3_ALERT_BAD_RECORD_MAC
E1124 11:01:22.214396573   10566 secure_endpoint.c:185]      Decryption error: TSI_DATA_CORRUPTED`

**Comment by WandyLau** [comment]
This error seems different from the above. And the different location in the source code. Maybe you can try my patch to see if it works.

**Comment by jboeuf** [comment]
@justinburke did you have the chance to look at the docker file?

**Comment by shimos** [comment]
@WandyLau Thank you for your fix! As a workaround, appending 0x00 to the root CA buffer seems to fix my issue.
So do you think basically this bug hits regardless of the concurrency of GRPC requests, whenever the contents of the memory just after the root CA buffer happens to be not zero?

**Comment by WandyLau** [comment]
@shimos yes, but It arise randomly. When more concurrency take place, there will be more chances that this issue will arise.

**Comment by vjpai** [comment]
@jboeuf : should this be re-assigned to you or someone on your team?

**Comment by danfairs** [comment]
I can confirm I'm seeing this too in Python with grpcio==1.10.0. It's getting trickier to work around now, since the official pubsub libraries are now threaded.

**Comment by danfairs** [comment]
I've also confirmed this is a problem in grpcio==1.8.2 (which is unfortunately the earliest version that core supports)

**Comment by danfairs** [comment]
Apologies - I've just noticed that this issue was specifically in the node bindings - I'll open another issue if that makes sense. In other news, I've rewound all the way back to 1.4.0, and I'm not getting the messages. When I have more time, I'll try to establish the exact version in which the problems start to occur.

**Comment by jiangtaoli2016** [comment]
We will have a PR on caching the X509_STORE on SSL root certificates (https://github.com/grpc/grpc/pull/14722). SSL default root certificates will be loaded only once. Let's see whether it will resolve.

**Comment by vjpai** [comment]
Following up: did merging #14722 resolve this issue>?

**Comment by vjpai** [comment]
Does anyone believe that this issue still exists? If not, I'd like to close it.
